### PR TITLE
Adding origin source locations to yul

### LINKED ISSUE #11086: [Yul] Syntax for "origin" source location
Since Yul's goal is to be used as an IR we want to keep the source locations of the *actual source code* in Yul. This is needed to have proper debugging support through the IR pipeline.

I could not find an issue for this, but we discussed it on several design meetings. If I recall correctly, the idea is that we change the `location` field in [Yul AST](https://github.com/ethereum/solidity/blob/develop/libyul/AST.h) to "annotations". This a could then contain Yul source location and "origin" source location. We could extend it with more debugging data to be transmitted later on.

For the syntax this is a [suggestion from 2017](https://gitter.im/ethereum/solidity-dev?at=601170c3428d9727dd38ea14) (ouch! -- well this link is from January, but based on an old commit) which is inspired by NatSpec:
```
/// @origin <sourcename>:<start>,<end>
```

Some name ideas: `@location`, `@origin`, `@generator`.

### PR DESCRIPTION
**Pull request by mijovic**
Fixes #11086

### CODE REVIEW

File: libyul/AST.h [hunk 1]
```diff
@@ -35,54 +35,63 @@ namespace solidity::yul
 
 using Type = YulString;
 
-struct TypedName { langutil::SourceLocation location; YulString name; Type type; };
+struct YulAstDebugData {
```
Comment (chriseth) [inline libyul/AST.h]: Just DebugData? (We are in solidity::yul namespace)

File: libyul/AST.h [hunk 2]
```diff
@@ -35,54 +35,63 @@ namespace solidity::yul
...
+	explicit YulAstDebugData(langutil::SourceLocation _yulLocation = {}, langutil::SourceLocation _sollLocation = {}):
+		yulLocation(std::move(_yulLocation)),
+		solLocation(std::move(_sollLocation))
+	{}
+	langutil::SourceLocation yulLocation;
+	langutil::SourceLocation solLocation;
```
Comment (chriseth) [inline libyul/AST.h]: irLocation and originLocation? sourceLocation? originalLocation?

Comment (mijovic) [reply libyul/AST.h]: went with `irLocation` and `originLocation`

File: libyul/AST.h (hunk 1 above)
Comment (mijovic) [reply libyul/AST.h]: Renamed to DebugData

File: test/cmdlineTests/exp_base_literal/output [hunk 3]
```diff
@@ -199,78 +199,157 @@ object "C_81" {
...
                 let _5 := var_a_4
                 let expr_24 := _5
+                /// @origin exp_base_literal/input.sol:196,200
                 let _6 := convert_t_rational_2_by_1_to_t_uint256(expr_23)
                 let expr_25 := checked_exp_t_rational_2_by_1_t_uint256(expr_24)
+                /// @origin exp_base_literal/input.sol:187,200
                 let var_w_22 := expr_25
+                /// @origin exp_base_literal/input.sol:212,219
+                /// @origin exp_base_literal/input.sol:212,216
+                /// @origin exp_base_literal/input.sol:214,215
```
Comment (mijovic) [inline test/cmdlineTests/exp_base_literal/output]: Do we want first location to be used, as it is the "longest"? I guess multiple calls to setLocation have happened before some code was generated...

File: test/cmdlineTests/strict_asm_jump/err [hunk 4]
```diff
@@ -3,4 +3,4 @@ Error: Function "jump" not found.
  --> strict_asm_jump/input.yul:1:3:
   |
 1 | { jump(1) }
-  |   ^^^^
+  |   ^^^^^^^
```
Comment (mijovic) [inline test/cmdlineTests/strict_asm_jump/err:6]: I am not sure from where this have been triggered...

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 5]
```diff
@@ -3167,6 +3167,23 @@ bool IRGeneratorForStatements::visit(TryCatchClause const& _clause)
 void IRGeneratorForStatements::setLocation(ASTNode const& _node)
 {
 	m_currentLocation = _node.location();
+	appendCurrentLocation();
+}
+
+void IRGeneratorForStatements::appendCurrentLocation()
+{
+	if (m_currentLocation.isValid() && m_code.tellp() != m_lastSourceLocationPosition)
```
Comment (mijovic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:3175]: not sure if using `tellp` is the best idea how to avoid multiple consecutive writes of source location, but it was the first idea I got.

**Comment by chriseth** [comment]
Please split this up further. The first PR should only be a change to AST.h

**Comment by mijovic** [comment]
> Please split this up further. The first PR should only be a change to AST.h

I wanted to do it as well. It is very hard to follow if I don't split.

File: libsolidity/analysis/ControlFlowBuilder.cpp [hunk 6]
```diff
@@ -482,8 +482,7 @@ void ControlFlowBuilder::operator()(yul::Identifier const& _identifier)
 			m_currentNode->variableOccurrences.emplace_back(
 				*declaration,
 				VariableOccurrence::Kind::Access,
-				_identifier.location
-			);
+				_identifier.debugData->irLocation);
```
Comment (axic) [inline libsolidity/analysis/ControlFlowBuilder.cpp:485]: Nit: I guess the `);` should stay on the new line.

File: libsolidity/analysis/ControlFlowBuilder.cpp [hunk 7]
```diff
@@ -390,7 +390,7 @@ bool ControlFlowBuilder::visit(InlineAssembly const& _inlineAssembly)
 void ControlFlowBuilder::visit(yul::Statement const& _statement)
 {
 	solAssert(m_currentNode && m_inlineAssembly, "");
-	m_currentNode->location = langutil::SourceLocation::smallestCovering(m_currentNode->location, locationOf(_statement));
+	m_currentNode->location = langutil::SourceLocation::smallestCovering(m_currentNode->location, locationOf(_statement)->irLocation);
```
Comment (axic) [inline libsolidity/analysis/ControlFlowBuilder.cpp:393]: Why isn't the `locationOf` helper just returning the location?

Comment (axic) [reply libsolidity/analysis/ControlFlowBuilder.cpp:393]: I think the helper should just be fixed, and all these places using the location should use `locationOf` or `irLocationOf` (a new helper).

File: libsolidity/analysis/TypeChecker.cpp [hunk 8]
```diff
@@ -743,7 +743,7 @@ bool TypeChecker::visit(InlineAssembly const& _inlineAssembly)
 			solAssert(var->type(), "Expected variable type!");
 			if (var->immutable())
 			{
-				m_errorReporter.typeError(3773_error, _identifier.location, "Assembly access to immutable variables is not supported.");
+				m_errorReporter.typeError(3773_error, _identifier.debugData->irLocation, "Assembly access to immutable variables is not supported.");
```
Comment (axic) [inline libsolidity/analysis/TypeChecker.cpp:746]: Thinking a bit more about this, I'm not 100% sold that the `visit(InlineAssembly)` section actually wants to use the `irLocation` and not the yul location.

The `irLocation` comes into play when creating the source mappings at the end of the IR pipeline.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 5 above)
Comment (axic) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:3175]: Is `setLocation` actually called so many repeated times where it should not be? I think we can just go with putting this out in `setLocation` unconditionally and call it a day.

File: libsolidity/parsing/Parser.cpp [hunk 9]
```diff
@@ -1302,7 +1302,7 @@ ASTPointer<InlineAssembly> Parser::parseInlineAssembly(ASTPointer<ASTString> con
 	if (block == nullptr)
 		BOOST_THROW_EXCEPTION(FatalError());
 
-	location.end = block->location.end;
+	location.end = block->debugData->irLocation.end;
```
Comment (axic) [inline libsolidity/parsing/Parser.cpp:1305]: I don't think this should be changed?

File: libyul/AST.h [hunk 10]
```diff
@@ -35,54 +35,62 @@ namespace solidity::yul
 
 using Type = YulString;
 
-struct TypedName { langutil::SourceLocation location; YulString name; Type type; };
+struct DebugData
+{
+	explicit DebugData(langutil::SourceLocation _yulLocation = {}, langutil::SourceLocation _sollLocation = {}): irLocation(std::move(_yulLocation)), originLocation(std::move(_sollLocation))
+	{}
+	langutil::SourceLocation irLocation;
```
Comment (axic) [inline libyul/AST.h:42]: Oh, perhaps this explains my confusion. I do not think this should be called `irLocation`, just kept as `location`.

File: libyul/AST.h [hunk 11]
```diff
@@ -35,54 +35,62 @@ namespace solidity::yul
...
+	template <class T> std::shared_ptr<DebugData> operator()(T const& _node) const
 	{
-		return _node.location;
+		return _node.debugData;
 	}
 };
 
 /// Extracts the source location from an inline assembly node.
-template <class T> inline langutil::SourceLocation locationOf(T const& _node)
+template <class T> inline std::shared_ptr<DebugData> locationOf(T const& _node)
```
Comment (axic) [inline libyul/AST.h:93]: I think this is a bad idea, this should stay at returning `langutil::SourceLocation` (and not the origin one).

**Comment by axic** [comment]
I think it would be really great to have a singular refactoring PR (which add DebugData without `originLocation` but does not make any behavioural changes). Then the second commit can deal with the IR changes.

We also want the parser to eventually read the `/// @origin` annotation and fill out the DebugData, but we should do that in a new PR.

**Comment by mijovic** [comment]
> I think it would be really great to have a singular refactoring PR (which add DebugData without `originLocation` but does not make any behavioural changes). Then the second commit can deal with the IR changes.
> 
> We also want the parser to eventually read the `/// @origin` annotation and fill out the DebugData, but we should do that in a new PR.

Thanks for suggestion, that sound like a good idea. Especially about the first PR that doesn't add any changes, except just refactoring to include DebugData instead of location.
Will try splitting it now and push only the refactor to this PR

**Comment by mijovic** [comment]
Closing because I opened new one #11294 that is pure refactoring. After that is finished, I'll add origin location in another PR
